include/hipSYCL/sycl/libkernel/sscp/builtins/core_typed.hpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * This file is part of AdaptiveCpp, an implementation of SYCL and C++ standard
+ * parallelism for CPUs and GPUs.
+ *
+ * Copyright The AdaptiveCpp Contributors
+ *
+ * AdaptiveCpp is released under the BSD 2-Clause "Simplified" License.
+ * See file LICENSE in the project root for full license details.
+ */
+// SPDX-License-Identifier: BSD-2-Clause
+#ifndef HIPSYCL_SSCP_BUILTINS_CORE_TYPED_HPP
+#define HIPSYCL_SSCP_BUILTINS_CORE_TYPED_HPP
+
+#include "builtin_config.hpp"
+
+#include <stddef.h>
+
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_local_id_x();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_local_id_y();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_local_id_z();
+
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_group_id_x();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_group_id_y();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_group_id_z();
+
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_local_size_x();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_local_size_y();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_local_size_z();
+
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_num_groups_x();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_num_groups_y();
+HIPSYCL_SSCP_BUILTIN __acpp_uint64 __acpp_sscp_get_num_groups_z();
```

Comment:
I don't understand what you mean. But I think the problem is that you have added an include to `core_typed.hpp` to `core.hpp`, so therefore, including `core.hpp` here causes an include cycle.

IMO, this indicates that there is a design flaw w.r.t what is defined in which includes. There should probably be 3 headers:
* `core.hpp`, only defining the core builtins. Can be included in bitcode libraries.
* `core_typed.hpp`, which includes `core.hpp` and defines the type-template function to calculate linear id. Can be included in bitcode libraries.
* `core_linear_id.hpp`, which includes `core_typed.hpp` and defines the `__acpp_sscp_get_*_linear_id<Dim>()` functions. Cannot be included in bitcode libraries.

Headers currently using `core.hpp` will then potentially also have to include the other headers. Outside of the bitcode libraries, we only include `core.hpp` in `sscp_backend.hpp` so that might not be too difficult.

Reply:
Okay, I think I understand. Implemented your suggestion